packages/viewer/src/components/viewer-pin-group/vertex-pin-group.tsx

```diff
@@ -0,0 +1,207 @@
+  @Prop({ mutable: true })
+  public elementBounds?: DOMRect;
+
+  /**
+   * The model that contains the entities and outcomes from performing pin operations
+   */
+  @Prop()
+  public pinModel: PinModel = new PinModel();
+
+  /**
+   * The controller that drives behavior for pin operations
+   */
+  @Prop()
+  public pinController?: PinController;
+
+  /**
+   * Whether or not the pin is "selected"
+   */
+  @Prop()
+  public selected = false;
+
+  @State()
+  private invalidateStateCounter = 0;
+
+  private labelEl: HTMLVertexViewerPinLabelElement | undefined;
+
+  protected componentDidLoad(): void {
+    this.setLabelObserver();
+  }
+
+  protected render(): JSX.Element {
+    if (this.pin == null) {
+      throw new Error('Unable to draw pin');
+    }
+    const computed = this.computeLabelLinePoint(this.pin);
+
+    const { pinPoint, labelPoint } = computed;
+
+    return (
+      <vertex-viewer-dom-group
+        id={`pin-group-${this.pin?.id}`}
+        data-testid={`pin-group-${this.pin.id}`}
+      >
+        <vertex-viewer-dom-element
+          key={`drawn-pin-${this.pin.id}`}
+          data-testid={`drawn-pin-${this.pin.id}`}
+          position={this.pin.worldPosition}
+          onPointerDown={(e) => {
+            e.stopPropagation();
+
+            this.pinController?.setSelectedPinId(this.pin?.id);
+          }}
+        >
+          {isTextPin(this.pin) && (
+            <div
+              id="pin-anchor"
+              class={classNames('pin-anchor', { selected: this.selected })}
+            ></div>
+          )}
+
+          {isDefaultPin(this.pin) && (
+            <vertex-viewer-icon
+              name="pin-fill"
+              size="lg"
+              class={classNames('pin', {
+                'pin-selected': this.selected,
+              })}
+            />
+          )}
+        </vertex-viewer-dom-element>
+
+        {isTextPin(this.pin) && (
+          <Fragment>
+            <vertex-viewer-pin-label-line
+              id={`pin-label-line-${this.pin?.id}`}
+              pinPoint={pinPoint}
+              labelPoint={labelPoint}
+            ></vertex-viewer-pin-label-line>
+
+            <vertex-viewer-pin-label
+              pin={this.pin}
+              ref={(el) => {
+                this.labelEl = el;
+              }}
+              elementBounds={this.elementBounds}
+              pinController={this.pinController}
+            ></vertex-viewer-pin-label>
+          </Fragment>
+        )}
+      </vertex-viewer-dom-group>
+    );
+  }
+
+  private invalidateState(): void {
+    this.invalidateStateCounter = this.invalidateStateCounter + 1;
+  }
+
+  private setLabelObserver(): void {
+    const label = this.labelEl?.addEventListener('labelChanged', () => {
+      this.invalidateState();
+    });
+
+    if (label != null) {
+      const resize = new ResizeObserver(() => this.invalidateState());
+
+      resize.observe(label);
+    }
+  }
+
+  private computeLabelLinePoint(pin: Pin): ComputedPoints {
```

Comment:
Wondering if we could break this out into a couple smaller functions. Looks like the first `screenPosition && pinPoint != null` is specific to text pins and the when just the `pinPoint` is returned it's a default pin - is that correct?

Might be nice if it were something like
```typescript
private computePinPoints(pin: Pin): ComputedPoints {
  if (this.elementBounds != null) {
    return isTextPin(this.pin) && this.pin.attributes.labelPoint != null ? 
      this.computeTextPinPoints(pin) : 
      this.computeDefaultPinPoints(pin);
  }

  return {
    pinPoint: pin.worldPosition,
  }
}
```